[Feature Request] `pathlib.Path` compatibility of index_file_path

## Proposal

Current use of path is well suited for `str` but not for ``pathlib.Path`

For instance:

```python
index_file_path = tdms_file + '_index'
```

could be adapted to:

```python
index_file_path = Path(Path(tdms_file).parent, Path(tdms_file).name + '_index')
```

I would be happy to suggest a PR on that issue

## Current undesired behavior

```python
tdms_file = PosixPath('/usr/src/app/data/my_tdms.tdms')

    def __init__(self, tdms_file):
        """ Initialise a new TdmsReader
    
        :param tdms_file: Either the path to the tdms file to read or an already
            opened file.
        """
        self._segments = None
        self._prev_segment_objects = {}
        self.object_metadata = OrderedDict()
        self._file_path = None
        self._index_file_path = None
    
        self._segment_channel_offsets = None
        self._segment_chunk_sizes = None
    
        if hasattr(tdms_file, "read"):
            # Is a file
            self._file = tdms_file
        else:
            # Is path to a file
            self._file = open(tdms_file, 'rb')
            self._file_path = tdms_file
>           index_file_path = tdms_file + '_index'
E           TypeError: unsupported operand type(s) for +: 'PosixPath' and 'str'
```

## Reference

https://docs.python.org/3/library/pathlib.html

--- a/nptdms/reader.py
+++ b/nptdms/reader.py
@@ -4,6 +4,10 @@
 import logging
 import os
 import numpy as np
+try:
+    from pathlib import Path
+except ImportError:
+    Path = None
 
 from nptdms import types
 from nptdms.common import ObjectPath, toc_properties
@@ -26,8 +30,8 @@ class TdmsReader(object):
     def __init__(self, tdms_file):
         """ Initialise a new TdmsReader
 
-        :param tdms_file: Either the path to the tdms file to read or an already
-            opened file.
+        :param tdms_file: Either the path to the tdms file to read
+            as a string or pathlib.Path, or an already opened file.
         """
         self._segments = None
         self._prev_segment_objects = {}
@@ -45,7 +49,13 @@ class TdmsReader(object):
             # Is path to a file
             self._file = open(tdms_file, 'rb')
             self._file_path = tdms_file
-            index_file_path = tdms_file + '_index'
+            if Path is not None:
+                # tdms_file may be pathlib.Path or string
+                file_path = Path(tdms_file)
+                index_file_path = Path(file_path.parent, file_path.name + '_index')
+            else:
+                # tdms_file must be a string
+                index_file_path = tdms_file + '_index'
             if os.path.isfile(index_file_path):
                 self._index_file_path = index_file_path
 

--- a/nptdms/reader.py
+++ b/nptdms/reader.py
@@ -4,10 +4,6 @@
 import logging
 import os
 import numpy as np
-try:
-    from pathlib import Path
-except ImportError:
-    Path = None
 
 from nptdms import types
 from nptdms.common import ObjectPath, toc_properties
@@ -47,15 +43,9 @@ class TdmsReader(object):
             self._file = tdms_file
         else:
             # Is path to a file
-            self._file = open(tdms_file, 'rb')
-            self._file_path = tdms_file
-            if Path is not None:
-                # tdms_file may be pathlib.Path or string
-                file_path = Path(tdms_file)
-                index_file_path = Path(file_path.parent, file_path.name + '_index')
-            else:
-                # tdms_file must be a string
-                index_file_path = tdms_file + '_index'
+            self._file_path = str(tdms_file)
+            self._file = open(self._file_path, 'rb')
+            index_file_path = self._file_path + '_index'
             if os.path.isfile(index_file_path):
                 self._index_file_path = index_file_path
 

--- a/nptdms/tdms.py
+++ b/nptdms/tdms.py
@@ -60,8 +60,8 @@ class TdmsFile(object):
     def read(file, memmap_dir=None):
         """ Creates a new TdmsFile object and reads all data in the file
 
-        :param file: Either the path to the TDMS file to read or an already
-            opened file.
+        :param file: Either the path to the tdms file to read
+            as a string or pathlib.Path, or an already opened file.
         :param memmap_dir: The directory to store memory mapped data files in,
             or None to read data into memory. The data files are created
             as temporary files and are deleted when the channel data is no
@@ -75,8 +75,8 @@ class TdmsFile(object):
         """ Creates a new TdmsFile object and reads metadata, leaving the file open
             to allow reading channel data
 
-        :param file: Either the path to the TDMS file to read or an already
-            opened file.
+        :param file: Either the path to the tdms file to read
+            as a string or pathlib.Path, or an already opened file.
         :param memmap_dir: The directory to store memory mapped data files in,
             or None to read data into memory. The data files are created
             as temporary files and are deleted when the channel data is no
@@ -89,16 +89,16 @@ class TdmsFile(object):
     def read_metadata(file):
         """ Creates a new TdmsFile object and only reads the metadata
 
-        :param file: Either the path to the TDMS file to read or an already
-            opened file.
+        :param file: Either the path to the tdms file to read
+            as a string or pathlib.Path, or an already opened file.
         """
         return TdmsFile(file, read_metadata_only=True)
 
     def __init__(self, file, memmap_dir=None, read_metadata_only=False, keep_open=False):
         """Initialise a new TdmsFile object
 
-        :param file: Either the path to the TDMS file to read or an already
-            opened file.
+        :param file: Either the path to the tdms file to read
+            as a string or pathlib.Path, or an already opened file.
         :param memmap_dir: The directory to store memory mapped data files in,
             or None to read data into memory. The data files are created
             as temporary files and are deleted when the channel data is no

--- a/nptdms/test/test_tdms_file.py
+++ b/nptdms/test/test_tdms_file.py
@@ -3,6 +3,7 @@
 from collections import defaultdict
 import logging
 import os
+import sys
 import tempfile
 from hypothesis import (assume, given, example, settings, strategies)
 import numpy as np
@@ -436,6 +437,23 @@ def test_read_with_index_file(test_file, expected_data):
     """ Test reading a file with an associated tdms_index file
     """
     with test_file.get_tempfile_with_index() as tdms_file_path:
+        tdms_file = TdmsFile.read(tdms_file_path)
+
+    for ((group, channel), expected_channel_data) in expected_data.items():
+        channel_obj = tdms_file[group][channel]
+        compare_arrays(channel_obj.data, expected_channel_data)
+
+
+@pytest.mark.skipif(sys.version_info < (3, 4), reason="pathlib only available in stdlib since 3.4")
+def test_read_file_passed_as_pathlib_path():
+    """ Test reading a file when using a pathlib Path object
+    """
+    import pathlib
+
+    test_file, expected_data = scenarios.single_segment_with_one_channel().values
+
+    with test_file.get_tempfile_with_index() as tdms_file_path_str:
+        tdms_file_path = pathlib.Path(tdms_file_path_str)
         tdms_file = TdmsFile.read(tdms_file_path)
 
     for ((group, channel), expected_channel_data) in expected_data.items():

